### DISCUSSION TITLE
Proposal: "Keptn Lighthouse" sub-project - generalizing Keptn SLI/SLO evaluations, and OpenSLO adoption

### DISCUSSION THREAD

**Discussion by oleg-nenashev** [keptn/keptn#8985] [Ideas]
There is an idea to generalize how Keptn handles SLIs and SLOs. We have a Keptn Lighthouse service that IMHO could provide standard interfaces for SLIs/SLOs (CloudEvents as now, OTel, OpenSLO) and allow users to evaluate SLOs in a generalized way while retrieving data through Keptn SLI providers which we have for Prometheus/Datadog/Dynatrace/SumoLogic at the moment.

I think such a solution could be used not only inside Keptn but also by many others in the cloud native ecosystem. It could also drive adoption for Keptn and Keptn Lifecycle Controller that could/would natively integrate with this new subproject.

![image](https://user-images.githubusercontent.com/3000480/195020667-aac1a97f-c930-44da-abc6-ef98d412757a.png)

Overview slides: https://docs.google.com/presentation/d/1Y_jSgWN6KM578IAJu-eBWtz7kj0c-JTWxS_XRs15uD4/edit?usp=drivesdk

Suggested plan:

- [ ] Start  a new repository in sandbox for docs/demos/tasks and interested parties
- [ ] TBD - start a working group
- [ ] Prototype Keptn LightHouse, based on the current Keptn Codebase, It would be a Helm chart with 2 services, and without a dedicated control plane. There might be API service too, TBD

Any feedback would be welcome!

**Comment by thisthat** [comment]
I support the idea! 👍 
Could you please elaborate more on which 2services you are thinking about when discussing the Helm chart? @oleg-nenashev

**Comment by vadasambar** [comment]
I support the idea too. 

1. What does `No support for expressions vs. the baseline` in [this slide](https://docs.google.com/presentation/d/1Y_jSgWN6KM578IAJu-eBWtz7kj0c-JTWxS_XRs15uD4/edit#slide=id.g1393f44ecf7_0_23) mean? Are we talking about comparison with previous SLO results that Keptn SLO supports but OpenSLO doesn't seem to support? 

Relevant section from Keptn SLO
![image](https://user-images.githubusercontent.com/34534103/195517731-b53c12e0-4dc9-4254-b178-e6118920e0e1.png)

**Reply by vadasambar** [reply]
P.S. 
2. What does `expressions` and `baseline` mean here?

**Comment by vadasambar** [comment]
What does `Actions` in  [`Actions for sending events`](https://docs.google.com/presentation/d/1Y_jSgWN6KM578IAJu-eBWtz7kj0c-JTWxS_XRs15uD4/edit#slide=id.g15a583b85ff_0_11) mean? 

What I understood from the that slide is, OpenSLO spec might have to support cloud events but I don't understand what `Actions` for sending events mean. Is this some kind of a trigger?

**Comment by oleg-nenashev** [comment]
Also @keptn/lifecycle-controller-maintainers implemented the CRD-driven SLI/SLO evaluation PoC in Keptn Lifecycle Controller. See in the bottom of the diagram below. The format is really primitive at the moment, just for the experiments. I believe bits of this architecture could be also reused in Keptn Lighthouse

![image](https://user-images.githubusercontent.com/3000480/197013853-d9ae8590-b394-47ea-86ef-febc1a331d24.png)

**Comment by vadasambar** [comment]
Started a discussion issue in OpenSLO: https://github.com/OpenSLO/OpenSLO/issues/187

**Comment by nieomylnieja** [comment]
hey there :wave: wanted to check If the discussion is still open on your end :) I wanted to get involved from the OpenSLO side of things, but don't want to burn time if the topic is dead
